src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/ExpressionBinder.cs

```diff
@@ -1135,18 +1135,8 @@ private Expr AdjustMemberObject(SymWithType swt, Expr pObject, out bool pfConstr
                     pfConstrained = true;
                 }
 
-                Expr objNew = tryConvert(pObject, swt.GetType(), CONVERTTYPE.NOUDC);
-
-                // This check ensures that we do not bind to methods in an outer class
-                // which are visible, but whose this pointer is of an incorrect type...
-                // ... also handles case of calling an pObject method on a RefAny value.
-                // WE don't give a great message for this, but it'll do.
-                if (objNew == null)
-                {
-                    throw ErrorContext.Error(ErrorCode.ERR_WrongNestedThis, swt.GetType(), pObject.Type);
-                }
-
-                pObject = objNew;
+                Debug.Assert(tryConvert(pObject, swt.GetType(), CONVERTTYPE.NOUDC) != null);
```

Comment:
This looks duplicated. Perhaps should just tryConvert and assert that the result is not null.

Reply:
The `isAssigned` and `pfConstrined` deal above seems strange. Hardly useful when everything is compiled into expression trees.